### PR TITLE
Fix double-scheduling of dc in HSS hopper matmuls

### PR DESCRIPTION
**Pull request by jacobhinkle**
When we do not have an epilogue (not even a cast), it might be the case that the original `MmaOp` has output which is a Fusion output. In this case the cached output which we often call `dc` is actually an `mma_result`. Currently this causes us to schedule that tensor once in `scheduleMmaResults` then again in `scheduleEpilogue`, leading to an esoteric error (see included test). This PR simply skips scheduling those tensors directly if they are already known to be mma results.

### CODE REVIEW

**Comment by jacobhinkle** [comment]
!test

**Comment by jacobhinkle** [comment]
We would have hit this eventually but the HSS tests are still guarded against Hopper. I'm posting this now to unblock some internal heuristics work.

**Review by rdspring1 — APPROVED** [review]

File: tests/cpp/test_matmul_scheduler.cpp [hunk 1]
```diff
@@ -2660,6 +2660,66 @@ TEST_F(MatmulSchedulerTest, SegmentMatmulOpUnsupportedDtype) {
...
+  MatMulTileOptions gemm_tile;
+  gemm_tile.cta_tile = GemmTile(128, 128, 16);
+  gemm_tile.warp_tile = GemmTile(64, 64, 16);
+  mparams.tile_sizes = gemm_tile;
+  mparams.async_gmem_load_operands = true;
+  mparams.circular_buffer_options.circular_buffer_smem_write = true;
+  mparams.circular_buffer_options.circular_buffer_smem_read = true;
+  mparams.circular_buffer_options.smem_circular_buffer_stage = 2;
+  // TODO: Currently we use stmatrix whenever this is true. We cannot do that
+  // when the dtype is not 16 bits.
```
Comment (jacobhinkle) [inline tests/cpp/test_matmul_scheduler.cpp:2709]: @protonu we need to handle all possible dtypes in the epilogue.

File: csrc/scheduler/hopper_multi_matmul.cpp
Comment (protonu) [inline csrc/scheduler/hopper_multi_matmul.cpp:1]: Are we still using `stmatrix` when the output is 32b?

File: csrc/scheduler/hopper_multi_matmul.cpp [hunk 2]
```diff
@@ -554,6 +562,14 @@ void HopperMultipleMatmulScheduler::scheduleEpilogue() {
...
       TensorView* d_smem = cacheAfter(dc, LoadStoreOpType::Set);
 
+      std::vector<TensorView*> tvs_to_schedule{d, d_smem};
+      if (std::find(mma_results_.begin(), mma_results_.end(), dc) ==
+          mma_results_.end()) {
+        // Skip scheduling dc if it is an mma_result. This can happen if we are
+        // not casting back to half-precision in the output
+        tvs_to_schedule.push_back(dc);
+      }
+
```
Comment (protonu) [inline csrc/scheduler/hopper_multi_matmul.cpp:572]: Are we still using `stmatrix` if the output is fp32 and there wasn't a cast?

Comment (jacobhinkle) [reply csrc/scheduler/hopper_multi_matmul.cpp:572]: Yes, if you enable `use_smem_epilogue` in the included test we hit an error in `scheduleStMatrixForMmaOutput`.

Comment (jacobhinkle) [reply csrc/scheduler/hopper_multi_matmul.cpp:572]: So we need an if/else here that checks the dtype of `d_smem` and schedules with vectorized stores instead if not 16bit https://github.com/NVIDIA/Fuser/blob/230f633eb995d6690ba67c57117c3b94ea26f25b/csrc/scheduler/hopper_multi_matmul.cpp#L584-L588

**Pull request merged**